param/__init__.py

```diff
@@ -1287,6 +1287,73 @@ def __init__(self, **params):
         super(Array,self).__init__(ndarray, allow_None=True, **params)
 
 
+class DataFrame(ClassSelector):
+    """
+    Parameter whose value is a pandas DataFrame.
+    """
+    __slots__ = ['rows','columns']
+
+    def __init__(self, rows=None, columns=None,**params):
+        from pandas import DataFrame as pdDFrame
+        self.rows = rows
+        self.columns = columns
+        super(DataFrame,self).__init__(pdDFrame, allow_None=True, **params)
+        self._check_value(self.default)
+
+    def _check_value(self,val,obj=None):
+        super(DataFrame, self)._check_value(val, obj)
+        length_error = 'Provided DataFrame has {found} columns when {expected} were expected'
+
+        if self.columns is None:
+            pass
+        elif isinstance(self.columns, list):
+            if len(val.columns) != len(self.columns):
+                raise Exception(length_error.format(found=len(val.columns), expected=len(self.columns)))
+            difference = set(val.columns) ^ set(self.columns)
+            if difference:
+                msg = 'Provided DataFrame columns {found} differ from expected columns {expected}'
+                raise Exception(msg.format(found=list(val.columns), expected=self.columns))
```

Comment:
I would argue for a different approach to validating DataFrame columns.  If someone provides a list of the columns they are expecting from the DataFrame, I think we should only validate that those columns are present.  E.g. if `['x','y']` is specified, it should be treated as a declaration that we will be accessing columns `x` and `y` from the provided dataframe, and so we ensure that those columns exist.  Conversely, if someone specifies cols to be `3`, it's a declaration that we will be accessing columns by number, and that there will need to be three of them, regardless of their names.  

For the numeric case it's not clear to me whether 3 should mean "at least 3", or "exactly 3", as there are arguments both ways. The reason for "at least 3" is that it's expressing specifically what we actually require, i.e. it's a declaration that we will access the first three columns by number; any additional columns won't hurt us but won't be accessed unless we first check that they exist. The reason for "exactly 3" would be to catch errors where someone supplies a larger one but doesn't realize that only the first 3 columns will be used. One could also imagine accepting a tuple to make it feasible to express precisely what you mean (like numeric bounds in Param in general, i.e.. with lower and upper bounds specified explicitly).  Unfortunately, a tuple like that could be ambiguous with the list of columns, e.g. when serialized to JSON (losing the distinction between tuples and lists), so I think I'd favor just picking whether a single numeric argument means "exactly" or "at least".  I think my vote is for "at least".

Reply:
The columns suggestion is now implemented in 3cf180b.

Reply:
Looks good, thanks.  @philippjfr?